---
title: "ukb_bfz"
author: "Innocent G Asiimwe"
date: "`r format(Sys.Date(), '%d %b %Y')`"
output:
  html_document: default
bibliography: references.bib
csl: frontiers-in-pharmacology.csl
---

<style type="text/css">
  body{
  font-size: 12pt;
}
</style>

```{r setup (Step 1), include = FALSE} 
knitr::opts_chunk$set(echo = TRUE) 

# Step 1: load libraries and relevant functions
# ---------------------------------------------
library(data.table) # reading tables faster
library(tidyverse) # data processing 
library(lubridate) # working with dates
library(mice)  # for imputation
library(qqman) # for generating Manhattan and QQ plots
library(HardyWeinberg) # for calculating Hardy-Weinberg equilibrium
library(gridExtra) # for arranging multiple ggplots on the same page
library(grid) # for visualization
library(psych) # for enhanced scatterplot
library(rms) # for assessing non-linearity using restricted cubic splines
library(pwr) # for sample size calculations
library(huxtable) # for generating tables
library(rvest) # for webscraping (getting genes and functional significance)
library(magick) # image annotation
library(rbioapi) # pathway analysis using Reactome
library(pdftools) # pdf manipulation
source("relevant_functions.R")
```

### <ins>**Genome-wide association studies of Bendroflumethiazide-related outcomes in the UK Biobank**</ins>

<br>

#### **Introduction**
* Thiazide diuretics, including Bendroflumethiazide (BFZ), are among the most widely-used, most effective and least costly therapies for hypertension. 
* Many thiazide-treated patients do not attain adequate blood pressure (BP) response while many experience electrolyte disorders that can result in adverse outcomes, including mortality. 
* `r ukb_colour_format("*We will therefore conduct genome-wide association studies (GWASs) using BFZ-treated patients in the UK Biobank to identify the genomic factors that influence BP response and the risk of two electrolyte disorders (hyperglycaemia and hyperuricaemia).*",'blue')`

<br>

\newpage
#### **Methods**
$\underline{Study\space design\space and\space setting}$

* UK Biobank, a population-based prospective cohort study [@RN1079].
  + Approximately half a million UK individuals, aged between 40–69 at recruitment that occurred between 2006–2010. 
  + Comprehensive genetic (>93 million markers) and phenotypic (including linked primary health care records) data. 

<br>

$\underline{Study\space population}$

* Prevalent and newly-initiated BFZ users who will be identified using data fields and coding algorithms provided by the UK Biobank.
  + Prevalent users (current analysis) identified using the UK Biobank **Data-Field 20003** (“Treatment/medication code” containing self-reported medications).
  + BFZ given singly or as part of a combination.

<br>

$\underline{Outcomes}$

* Co-primary/efficacy outcomes: 
  + Systolic BP (automated, **Data-Field 4080**; manual, **Data-Field 93**)
  + Diastolic BP (automated, **Data-Field 4079**; manual, **Data-Field 94**) 
* Secondary safety/electrolyte disorder outcomes: 
  + Blood glucose (<b>Data-Field 30740</b>)
  + Serum urate (<b>Data-Field 30880</b>)
  + Hyponatraemia (serum sodium) and hypokalaemia (serum potassium) not measured at baseline. We may later be able to include them using primary care records.
* Current analysis is continuous outcomes, to later consider dichotomous outcomes.

<br>

$\underline{GWAS\space covariates}$

* First ten principal components of genetic ancestry (**Data-Field 22009**)
* Non-genetic covariates:
  + Age at baseline (years, <b>Data-Field 21003</b>)
  + Sex (male vs female, **Data-Field 31**)
  + Body mass index (Kg/m^2^, **Data-Field 21001**)
  + Smoking status (current vs former vs never, **Data-Field 20116**)
  + Alcohol drinker status (current vs former vs never, **Data-Field 20117**)
  + Degree of physical activity (Types of physical activity in last 4 weeks, coded to heavy [a) strenuous sports and b) heavy DIY like weeding, lawn mowing, carpentry, digging], light [a) other exercises like swimming, cycling, keep fit, bowling, b) light DIY like pruning, watering the lawn, and c) walking for pleasure (not as a means of transport) and none, **Data-Field 6164**)
  + Townsend index reflecting socioeconomic status (a continuous score, **Data-Field 189**)

<br>

$\underline{Genotyping\space and\space QC}$

* Two similar (about 95% shared marker content) arrays: 
  + UK BiLEVE Axiom Array by Affymetrix (49 950 participants, 807 411 markers)
  + Applied Biosystems UK Biobank Axiom Array (438 427 participants, 825 927 markers)
  + Phasing (SHAPEIT3) and imputation (IMPUTE4) using UK10K, 1000 Genomes phase 3, and Haplotype Reference Consortium panels increased the number of markers to > 93 million.
* Per-SNP QC: SNPs included if MAF ≥ 1%, missingness ≤ 5%, HWE > 1E-06, and info score > 0.4.
* Per-sample QC: removed participants who:
  + Were outliers based on missing rate (>5%) and heterozygosity adjusted for population structure,
  + Had mismatches between self-reported and marker-inferred sex, and/or,
  + Had greater than 3rd-degree relatedness with another included participant.

<br>

$\underline{Minimum\space sample\space size}$

* Based on the primary outcomes of systolic and diastolic BPs.
* Assuming a MAF of 20%, 80% power and a significance threshold of 5 × 10-8:
  + Systolic BP: `r ukb_colour_format(ukb_min_sample(10, 19),'blue')` participants, standard deviation (SD) of 19 mmHg for 475 540 UK Biobank participants and effect size 10 mmHg (obtained from previous studies)
  + Diastolic BP: `r ukb_colour_format(ukb_min_sample(5, 11),'blue')` participants, SD of 11 mmHg (UK Biobank) and effect size 5 mmHg (previous studies)
* These sample sizes achievable for Whites. Due to smaller sample sizes, the proposal is to use other races (Blacks and Asians) for replication.

<br>

$\underline{Statistical\space analysis}$

* Outcome transformation: log transformation for continuous outcomes.
* Predictor handling
  + Quantitative predictor variables neither transformed nor categorized. 
  + Nonlinearity between continuous predictors and continuous outcomes assessed using restricted cubic splines in the rms R package.
  + Categorical variables dummy coded in R.
* Missing data
  + Participants with missing outcome data excluded from the corresponding analyses.
  + No participant excluded on the basis of missing covariate data. Missing genotype data imputed using IMPUTE4, missing non-genetic covariates imputed using single imputation (MICE R package).
* Multivariable linear regression
  + Additive mode of inheritance assumption
  + Adjustment for the seven non-genetic covariates and the first 10 principal components of genetic ancestry.
  + Two significance thresholds: a genome-wide statistical significance threshold of p < 5 × 10-8, as well as a nominal significance threshold of p < 1 × 10-5. 
  + No further adjustment for multiplicity testing as this analysis is exploratory. 
  + All analyses will be undertaken using SNPTEST Version 2. 
* Further analysis using the Single Nucleotide Polymorphism (dbSNP) database, the Genotype-Tissue-Expression (GTEx) portal, Reactome (pathway analysis), LocusZoom and Haploview.

<br>

\newpage
#### **Results**

```{r Step 2, include = FALSE, eval = FALSE} 
#Step 2: Turn into long-format for Treatment/medication code (we need to subset 
#the population first for speed and space downstream)
#-------------------------------------------------------------------------------
dict <- as_tibble(fread("Data_Dictionary_Showcase.csv")) # import dictionary
bd <- as_tibble(fread("bd_enc.csv")) # load UK biobank dataset
bd_BFZ <- select(bd, contains(c("f.eid", "f.53.", "f.20003.")))
bd_BFZ_long <- ukb_reshape_long(bd_BFZ, dict) 

dict_meds <- as_tibble(fread("Codings.csv")) #codings for medications == 4
dict_meds <- dict_meds %>%
  subset(Coding == 4) %>%
  select(Value, Meaning)

colnames(dict_meds) <- c("Treatment/medication code","Treatment/medication")
dict_meds$`Treatment/medication code` <-
  parse_integer(dict_meds$`Treatment/medication code`)
  #change to numeric to ensure compatibility with bd_BFZ_long

bd_BFZ_long <- bd_BFZ_long %>%
  left_join(dict_meds) %>%
  select(!`Treatment/medication code`) %>%
  filter(I == ".0.") # Join data set with the dict and retain only first instance

UKBB_participants <- length(unique(bd_BFZ_long$f.eid)) - length(withdrawn_consent)
```

```{r Step 3, include = FALSE, eval = FALSE}
# Step 3: Identify BFZ users 
# ---------------------------
withdrawn_consent <- as_tibble(fread("withdrawn_consent.csv")) # this is emailed to
# approved users or can be accessed through the UK Biobank Access Management System.

drug <- "atenolol+bendrofluazide|atenolol+bendroflumethiazide|bendrofluazide|bendroflumethiazide|bzt - bendrofluazide|neo-naclex 5mg tablet|bendrofluazide+potassium 2.5mg/7.7mmol m/r tablet|bendroflumethiazide+potassium 2.5mg/7.7mmol m/r tablet|neo-naclex k m/r tablet|nadolol+bendrofluazide 40mg/5mg tablet|nadolol+bendroflumethiazide 40mg/5mg tablet|propranolol hydrochloride+bendrofluazide 80mg/2.5mg capsule|prestim tablet|timolol maleate+bendrofluazide 10mg/2.5mg tablet|timolol maleate+bendroflumethiazide 10mg/2.5mg tablet|nadolol+bendrofluazide 80mg/5mg tablet|timolol maleate+bendrofluazide 20mg/5mg tablet”, and “neo-bendromax 2.5mg tablet"

index <- str_detect(tolower(bd_BFZ_long$`Treatment/medication`), drug)
index[is.na(index)] <- FALSE
bd_BFZ <- bd_BFZ_long[index, ] %>%
  select(!`Treatment/medication`) %>%
  unique() %>%
  filter(!f.eid %in% withdrawn_consent$V1)

BFZ_participants <- nrow(bd_BFZ)
```

```{r Step 4, include = FALSE, eval = FALSE}
# Step 4: Identify BFZ users with genetic data (exclude outliers and relatives)
# -----------------------------------------------------------------------------
genotype_data <- as_tibble(fread("with_genotype_data.csv")) # list of participants
# with genetic data can be obtained from the .fam files
outliers <- as_tibble(fread("outliers.csv")) %>%
  filter(is.na(f.22027.0.0) == FALSE)

bd_BFZ <- bd_BFZ %>%
  filter(f.eid %in% genotype_data$feid) %>%
  filter(!f.eid %in% outliers$f.eid) 

with_genetic <- nrow(bd_BFZ)

# To remove participants who are related, missingness and realatedness .csv files
# are required, and a script could be run as shown below:
# missingness <- as_tibble(fread("missingness.csv"))
# related <- as_tibble(fread("UKBB_relatedness.csv")) %>%
#   filter(Kinship > 0.0884) %>%
#   select(ID1, ID2) %>%
#   mutate(ID1 = as.character(ID1), ID2 = as.character(ID2)) %>%
#   group_by(ID1) %>%
#   dplyr::summarize(include = ukb_related(., missingness, bd_BFZ))
# However, the above takes too much time, and it doesn't cater for multi-way relations.

# An example of a multi-way relation is a participant (x) being a child of (y) and (z).
# (x) is related to both (y) and (z), but (y) and (z) may be unrelated.
# In this case, it is more appropriate to exclude (x) and retain (y) and (z) as 
# this means a higher analyzed sample size, even without considering the amount of
# missingness i.e. (x) should be excluded even if (x) has the lowest missingness.

missingness <- as_tibble(fread("missingness.csv"))
related <- as_tibble(fread("UKBB_relatedness.csv")) %>%
  filter(Kinship > 0.0884) %>%
  select(ID1, ID2)

related_ids <- unique(c(related$ID1, related$ID2))
related_ids <- related_ids[related_ids %in% bd_BFZ$f.eid] # limit to only those ids
# present in the bd_BFZ dataframe.

related.df <- tibble(ID1 = vector("integer", length(related_ids)),
                     include = vector("character", length(related_ids))
                     )

for(i in seq_along(related_ids)) {
  id <- related_ids[[i]]
  id_df <- related %>% 
    filter(ID1 == id | ID2 == id)
  all_ids <- id_df %>%
    as.matrix() %>%
    as.integer()
  id_df_3_way <- related %>%
    filter(ID1 %in% all_ids | ID2 %in% all_ids) # this should capture 2+-way relations
  all_IDS <- id_df_3_way %>%
    as.matrix() %>%
    as.integer()
  id_df_3_way <- related %>%
    filter(ID1 %in% all_IDS | ID2 %in% all_IDS) # additional iteration in case of multiple relations
  id_count <- id_df_3_way %>%
    as.matrix() %>%
    as.integer() %>%
    as_tibble() %>%
    count(value)
  related.df[i, 1] <- id
  if(nrow(id_df_3_way) == 1) {  # one pair of related individuals
    related.df[i, 2] <- ukb_related(id_df_3_way, missingness, bd_BFZ)
  } else if(sum(id_count$n) == 4){ # for this 3-way, sum of counts is 4 (1 + 1 + 2)
    id_df_n_way <- tibble(ID1 = id_count$value[id_count$n == 1][1],
                          ID2 = id_count$value[id_count$n == 1][2]) # to maximize
    # sample size, remove the relation that is common to both
    in_bd_BFZ <- bd_BFZ %>%
      filter(f.eid %in% as.integer(as.matrix(id_df_n_way))) %>%
      nrow() # to check if the two values are present
    if (in_bd_BFZ < 2) {
      ids_BFZ <- id_df_3_way %>%
        as.matrix() %>%
        as.integer() %>%
        unique()
      ids_BFZ <- ids_BFZ[ids_BFZ %in% bd_BFZ$f.eid]
      id_df_n_way <- tibble(ID1 = ids_BFZ[1],
                            ID2 = ids_BFZ[length(ids_BFZ)])
    }
    related.df[i, 2] <- ukb_related(id_df_n_way, missingness, bd_BFZ)
  } else if(nrow(id_count) == 3 && sum(id_count$n) == 6){ # for this 3-way,
    # sum of counts is 6 (2 + 2 + 2)
    related.df[i, 2] <- ukb_related(id_df_3_way, missingness, bd_BFZ)
    # just chose the least missing id of the 3
  } else { # any other relations
    in_bd_BFZ <- bd_BFZ %>%
      filter(f.eid %in% as.integer(as.matrix(id_df_3_way))) %>%
      select(f.eid)
    id_df_n_way <- tibble(ID1 = in_bd_BFZ$f.eid[1],
                          ID2 = in_bd_BFZ$f.eid[nrow(in_bd_BFZ)])
    related.df[i, 2] <- ukb_related(id_df_n_way, missingness, bd_BFZ)
  }
  message(paste(round(i / length(related_ids) * 100, 3), "% complete", sep = ""))
}

for_inclusion <- related.df %>%
  filter(is.na(include) == FALSE) %>%
  unique()
for_inclusion <- unique(for_inclusion$include)

for_exclusion <- related_ids[!related_ids %in% for_inclusion] 

bd_BFZ <- bd_BFZ %>%
  filter(!f.eid %in% for_exclusion)
unrelated <- nrow(bd_BFZ) # unrelated dataset
```

```{r Step 5, include = FALSE, eval = FALSE}
# Step 5: Sub-set the bd data frame and convert it to a long format (remove 
# medications as already obtained, for speed and also not needed - we know these 
# patients took BFZ at baseline)
# Also remove genetic covariates, which will be added separately
# ------------------------------------------------------------------------------
bd_long <- bd %>%
  filter(f.eid %in% bd_BFZ$f.eid) %>%
  select(!contains(c("f.20003.", "f.22009."))) %>%
  ukb_reshape_long(dict) %>%
  filter(I == ".0.")
```

```{r Step 6, include = FALSE, eval = FALSE}
# Step 6: Get covariates data frame
# ---------------------------------
# First sort out physical exercise
pa_levels <- c("Strenuous sports",
               "Heavy DIY (eg: weeding, lawn mowing, carpentry, digging)",
               "Other exercises (eg: swimming, cycling, keep fit, bowling)",
               "Light DIY (eg: pruning, watering the lawn)",
               "Walking for pleasure (not as a means of transport)",
               "None of the above", "Prefer not to answer")
bd_pa <- bd_long %>%
  mutate (pa = factor(`Types of physical activity in last 4 weeks`,
                      levels = pa_levels)) %>%
  select(f.eid, I, `Date of attending assessment centre`, pa) %>%
  group_by(f.eid) %>%
  summarise(pa = pa_levels[min(as.numeric(pa))])

# Sort out other covariates
bd_cov <- bd_long %>%
  select(f.eid, I, `Date of attending assessment centre`,
         `Age when attended assessment centre`,
         Sex, `Body mass index (BMI)`, `Smoking status`,
         `Alcohol drinker status`,
         `Townsend deprivation index at recruitment`, `Ethnic background`) %>%
  unique() %>%
  rename(date = `Date of attending assessment centre`,
         age = `Age when attended assessment centre`,
         sex = Sex, bmi = `Body mass index (BMI)`,
         smoking_status = `Smoking status`,
         alcohol_status = `Alcohol drinker status`,
         townsend_index = `Townsend deprivation index at recruitment`,
         race = `Ethnic background`)

# Combine with PA
bd_cov <- left_join(bd_cov, bd_pa)

# Deal with factors
bd_cov <- bd_cov %>%
  mutate(pa = fct_collapse(pa,
         heavy = c("Strenuous sports",
                    "Heavy DIY (eg: weeding, lawn mowing, carpentry, digging)"),
         light = c("Other exercises (eg: swimming, cycling, keep fit, bowling)",
                    "Light DIY (eg: pruning, watering the lawn)",
                    "Walking for pleasure (not as a means of transport)"),
         none = c("None of the above")
         ))
bd_cov <- bd_cov %>%
  mutate(race = fct_collapse(factor(race),
                             White = c("British", "Irish", "White",
                                       "Any other white background"),
                             Black = c("African", "Caribbean",
                                       "Any other Black background"),
                             Asian = c("Chinese", "Bangladeshi", "Indian",
                                       "Pakistani",
                                       "Any other Asian background"),
                             Mixed_other_unknown = c("Mixed",
                                                  "Other ethnic group",
                                                  "White and Asian",
                                                  "White and Black African",
                                                  "White and Black Caribbean",
                                                  "Any other mixed background",
                                                  "Do not know")
                             )
         )
factors <- c("sex", "smoking_status", "alcohol_status", "pa", "race")
levels <- list(c("Female", "Male"),
               c("Never", "Previous", "Current", "Prefer not to answer"),
               c("Never", "Previous", "Current", "Prefer not to answer"),
               c("none", "light", "heavy", "Prefer not to answer"),
               c("White", "Black", "Asian", "Mixed_other_unknown",
                 "Prefer not to answer")
               )
bd_cov[factors] <- bd_cov %>%
  select(all_of(factors)) %>%
  map2(levels, factor) %>%
  ukb_recode_factor_na

# Add the first 10 principal components of genetic ancestry
pcs <- as_tibble(fread("UKBB_principal_components.csv")) #load data
pcs <- pcs %>% filter(f.eid %in% bd_cov$f.eid)
colnames(pcs) <- c("f.eid", "C1", "C2", "C3", "C4", "C5", "C6", "C7", "C8",
                   "C9", "C10")
bd_cov <- left_join(bd_cov, pcs)
```

```{r Step 7, include = FALSE, eval = FALSE}
# Step 7: Extract outcomes (use manual bps when automated ones are missing), log
# transform at the end
# ------------------------------------------------------------------------------
bd_outcomes <- bd_long %>%
  select(f.eid, I, `Date of attending assessment centre`,
         `Systolic blood pressure, manual reading`,
         `Diastolic blood pressure, manual reading`,
         `Diastolic blood pressure, automated reading`,
         `Systolic blood pressure, automated reading`,
         Glucose, Urate) %>%
  unique() %>%
  rename(date = `Date of attending assessment centre`,
         sbp_m = `Systolic blood pressure, manual reading`,
         dpb_m = `Diastolic blood pressure, manual reading`,
         dpb_a = `Diastolic blood pressure, automated reading`,
         spb_a = `Systolic blood pressure, automated reading`,
         glucose = Glucose,
         urate = Urate)  %>%
  group_by(f.eid) %>%
  mutate_at(vars(sbp_m:urate), median, na.rm = TRUE) %>%
  unique() %>%
  mutate(sbp = ifelse(is.na(spb_a) == TRUE, sbp_m, spb_a),
                       dbp = ifelse(is.na(dpb_a) == TRUE, dpb_m, dpb_a)
         ) %>%
  select(-sbp_m, -dpb_m, -dpb_a, -spb_a) %>%
  mutate_at(vars(glucose:dbp), log)

bd_BFZ_df <- left_join(bd_cov, bd_outcomes)
write_rds(bd_BFZ_df, "bd_BFZ_df.rds") # save for downstream work

bd_BFZ_df <- read_rds("bd_BFZ_df.rds")
analyzed <- bd_BFZ_df %>%
  filter(race == "White" | race == "Black" | race == "Asian") %>%
  nrow()

analyzed_white <- ukb_subset_count(bd_BFZ_df, "White")
analyzed_white_bp <- ukb_subset_count(bd_BFZ_df, "White", "sbp")
analyzed_white_glucose <- ukb_subset_count(bd_BFZ_df, "White", "glucose")
analyzed_white_urate <- ukb_subset_count(bd_BFZ_df, "White", "urate")

analyzed_black <- ukb_subset_count(bd_BFZ_df, "Black")
analyzed_black_bp <- ukb_subset_count(bd_BFZ_df, "Black", "sbp")
analyzed_black_glucose <- ukb_subset_count(bd_BFZ_df, "Black", "glucose")
analyzed_black_urate <- ukb_subset_count(bd_BFZ_df, "Black", "urate")

analyzed_asian <- ukb_subset_count(bd_BFZ_df, "Asian")
analyzed_asian_bp <- ukb_subset_count(bd_BFZ_df, "Asian", "sbp")
analyzed_asian_glucose <- ukb_subset_count(bd_BFZ_df, "Asian", "glucose")
analyzed_asian_urate <- ukb_subset_count(bd_BFZ_df, "Asian", "urate")
```

<br>

```{r Figure 1, echo = FALSE, eval = FALSE}
DiagrammeR::grViz("
digraph graph1 {

# graph, node, and edge definitions
  graph [compound = true, nodesep = .5, ranksep = .25,
         color = red, layout = dot, rankdir = TB]

  node [fontname = Helvetica, fontcolor = black, fontsize = 12,
        shape = rectangular, fixedwidth = true, width = 3, 
        color = black, penwidth = 0.8]

  edge [color = blue, arrowhead = normal, arrowtail = none,
       arrowsize = 0.5, penwidth = 0.8]
   #a [label = '@@1 UK Biobank participants aged<br/> between 40–69 y at recruitment']
  a [label = < <b>@@1</b>  UK Biobank participants aged<br/> between 40–69 y at recruitment (2006–<br/>2010) and who had not withdrawn<br/> consent as of 22 August 2022>]
  b [label = < <b>@@2</b>  participants taking<br/> Bendroflumethiazide (BFZ) at<br/> recruitment>]
  c [label = < <b>@@3</b>  BFZ-treated participants with<br/> genetic data>]
  d [label = < <b>@@4</b>  unrelated BFZ-treated<br/> participants with genetic data>]
  e [label = < <b>@@5</b>  BFZ-treated participants<br/> included in analysis>]
  
  node [width = 1, fontsize = 10]
  f [label = < <I> <U> <b>@@6</b>  Whites<br/> </U> </I>
Blood pressure outcomes: <b>@@7</b><br ALIGN='LEFT'/>
   Blood glucose: <b>@@8</b><br ALIGN='LEFT'/>
   Serum urate: <b>@@9</b><br ALIGN='LEFT'/>>]
  g [label = < <I> <U> <b>@@10</b>  Blacks<br/> </U> </I>
Blood pressure outcomes: <b>@@11</b><br ALIGN='LEFT'/>
   Blood glucose: <b>@@12</b><br ALIGN='LEFT'/>
   Serum urate: <b>@@13</b><br ALIGN='LEFT'/>>]
  h [label = < <I> <U> <b>@@14</b>  Asians<br/> </U> </I>
Blood pressure outcomes: <b>@@15</b><br ALIGN='LEFT'/>
   Blood glucose: <b>@@16</b><br ALIGN='LEFT'/>
   Serum urate: <b>@@17</b><br ALIGN='LEFT'/>>]
  
  a -> b -> c -> d -> e -> {f g h}
  
  node [style = dashed, rankdir = TB, width = 2.7, fontsize = 12]
  i [label = < <b>@@18</b>  participants not taking<br/> Bendroflumethiazide at recruitment>]
  j [label = < <b>@@19</b>  BFZ-treated participants with no<br/> genetic data>]
  k [label = < <b>@@20</b>  BFZ-treated participants with<br/> greater than 3rd-degree relatedness<br/> with another included participant>]
  l [label = < <b>@@21</b>  BFZ-treated participants with<br/> mixed, other or unknown race>]
  
  edge [style = invis]
  i -> j -> k -> l
  
  edge [style = dashed]
  a -> i b -> j c -> k d -> l
}
  [1]: ukb_comma(UKBB_participants)
  [2]: ukb_comma(BFZ_participants)
  [3]: ukb_comma(with_genetic)
  [4]: ukb_comma(unrelated)
  [5]: ukb_comma(analyzed)
  [6]: ukb_comma(analyzed_white)
  [7]: ukb_comma(analyzed_white_bp)
  [8]: ukb_comma(analyzed_white_glucose)
  [9]: ukb_comma(analyzed_white_urate)
  [10]: ukb_comma(analyzed_black)
  [11]: ukb_comma(analyzed_black_bp)
  [12]: ukb_comma(analyzed_black_glucose)
  [13]: ukb_comma(analyzed_black_urate)
  [14]: ukb_comma(analyzed_asian)
  [15]: ukb_comma(analyzed_asian_bp)
  [16]: ukb_comma(analyzed_asian_glucose)
  [17]: ukb_comma(analyzed_asian_urate)
  [18]: ukb_comma(UKBB_participants - BFZ_participants)
  [19]: ukb_comma(BFZ_participants - with_genetic)
  [20]: ukb_comma(with_genetic - unrelated)
  [21]: ukb_comma(unrelated - analyzed)
  ",
  height = 400)
```
**Figure 1. Participant inclusion flow chart.** to be inserted here.

```{r Step 8, include = FALSE, eval = FALSE}
# Step 8: Descriptive analysis (tables, figures) etc, each race separately
#-------------------------------------------------------------------------
```

<br>

```{r Figure 2, include = FALSE, eval = FALSE}
# Box plots for continuous outcomes
bd_BFZ_df <- read_rds("bd_BFZ_df.rds")
bd_BFZ <- bd_BFZ_df %>%
  select(age:pa, glucose:dbp) %>%
  mutate(glucose = exp(glucose),
         urate = exp(urate),
         sbp = exp(sbp),
         dbp = exp(dbp)
         ) %>%
  filter(race == "White" | race == "Black" | race == "Asian" )

new_covariates <- bd_BFZ %>% keep(is.numeric) %>% colnames()
output <- list("list", length(new_covariates))
for(i in seq_along(new_covariates)){
  output[[i]] <- ukb_box_plots(bd_BFZ, new_covariates[[i]])
}
png("continuous_variables.png", width=1500, height=1500, res=120)
n_columns <- 3
grid.arrange(grobs = output, 
             ncol = n_columns)
dev.off()
```

**Figure 2. Participant characteristics (continuous independent covariates plus outcomes).** to be inserted here.

<br>

```{r Figure 3, include = FALSE, eval = FALSE}
# categorical variable plots
new_covariates <- bd_BFZ %>% keep(is.factor) %>% select(-race) %>% colnames()
output <- list("list", length(new_covariates))
for(i in seq_along(new_covariates)){
  output[[i]] <- ukb_bar_plots(bd_BFZ, new_covariates[[i]])
}
png("categorical_variables.png", width=1500, height=1500, res=120)
n_columns <- 2
grid.arrange(grobs = output, 
             ncol = n_columns)
dev.off()
```
**Figure 3. Participant characteristics (categorical independent covariates).** to be inserted here.

<br>

```{r Table 1, echo = FALSE, message = FALSE, eval = FALSE}
# With non-imputed data
bd_BFZ_df <- read_rds("bd_BFZ_df.rds")
races <- c("White", "Black", "Asian")
for(i in seq_along(races)){
  race_BFZ <- races[[i]]
  bd_BFZ_race <- bd_BFZ_df %>% 
    filter(race == race_BFZ) %>%
    select(age:townsend_index, pa, sbp, dbp, glucose, urate) %>%
    mutate(sbp = exp(sbp),
           dbp = exp(dbp),
           glucose = exp(glucose),
           urate = exp(urate)
    )
  # message("There are ", nrow(bd_BFZ_race), " ", race_BFZ, " patients")
  race_summary <- ukb_summary(bd_BFZ_race)
  colnames(race_summary) <- c("Independent variables", "statistic", race_BFZ)
  # write.csv(race_summary, 
  #           file = paste(race_BFZ, "_", "summary.csv", sep = ""), 
  #           row.names=FALSE)
  if (i == 1) race_summary_all <- race_summary 
  else race_summary_all <- full_join(race_summary_all, race_summary)
}
#race_summary_all <- race_summary_all %>% select(-statistic) 

covariate_list <- c("age", "sex", "bmi", "smoking_status", "alcohol_status", "townsend_index",
                    "pa", "sbp", "dbp", "glucose", "urate")
covariate_list_2 <- c("<i>Age (years)</i>", "<i>Sex, n (%)</i>", 
                      "<i>Body mass index (Kg/m<sup>2</sup>)</i>",
                      "<i>Smoking status, n (%)</i>", "<i>Alcohol status, n (%)</i>", 
                      "<i>Townsend index score</i>", 
                      "<i>Types of physical activity in last 4 weeks, n (%)</i>", 
                      "<i>Systolic blood pressure (mmHg)</i>",
                      "<i>Diastolic blood pressure (mmHg)</i>",
                      "<i>Blood glucose (mmol/L)</i>", "<i>Serum urate (umol/L)</i>"
                      )
hux_table_1 <- ukb_hux(race_summary_all, covariate_list, covariate_list_2) %>%
  set_font_size(9.8) %>%
  set_col_width(1, 0.13) %>%
  set_col_width(2, 0.16) %>%
  set_caption("<span style='color: black; font-weight: bold; font-size: 12pt'>Table 1: Participant characteristics<sup>a</sup></span>") %>%
  set_caption_pos("topleft") %>%
  add_footnote("<sup>a</sup>Less than 0.5% of the participants were excluded when analysing BP outcomes. The number of participants excluded for blood glucose (about 13%) and serum urate (about 5%) was higher. A comparison of those included vs excluded in the analysis for these two outcomes is shown in the next slide.", font_size = 8.5) %>%
  set_escape_contents(col = 1, value = FALSE) %>%
  set_escape_contents(row = 1, value = FALSE) # allows you to include raw HTML
hux_table_1
```

```{r Table 1b, include = FALSE, eval = FALSE}
hux_table_1 <- hux_table_1 %>%
  set_width(0.9)  
quick_html(hux_table_1, file = "hux_table_1.html", open = FALSE)
webshot2::webshot(url = "hux_table_1.html", file = "hux_table_1.png", zoom = 5)
```
**Table 1** to be inserted here.

<br>

```{r Table 2, echo = FALSE, message = FALSE, eval = FALSE}
races <- c("White", "Black", "Asian")
outcomes <- c("glucose", "urate")
race_summary_outcomes <- vector("list", length(outcomes))
names(race_summary_outcomes) <- outcomes
nrow_included_outcomes <- vector("list", length(outcomes))
names(nrow_included_outcomes) <- outcomes
nrow_excluded_outcomes <- vector("list", length(outcomes)) # for both outcomes
names(nrow_excluded_outcomes) <- outcomes
# Use imputed dataset
for(j in seq_along(outcomes)){
  outcome <- outcomes[[j]]
  nrow_included <- vector("double", length(races))
  nrow_excluded <- vector("double", length(races))
  for(i in seq_along(races)){
    race_BFZ <- races[[i]]
    bd_imp <- read_rds(paste(race_BFZ, "_bd_imp.rds", sep ="")) %>%
      select(age:pa)
    bd_no_imp <- read_rds("bd_BFZ_df.rds") %>% 
      filter(race == race_BFZ) %>%
      select(all_of(outcomes))
    bd <- bind_cols(bd_imp, bd_no_imp)
    bd_outcome <- bd %>%
      select(age:pa, all_of(outcome))
    included <- bd_outcome[!is.na(bd_outcome[ncol(bd_outcome)]),] %>%
      select(-all_of(outcome))
    excluded <- bd_outcome[is.na(bd_outcome[ncol(bd_outcome)]),] %>%
      select(-all_of(outcome))
    nrow_included[[i]] <- nrow(included)
    nrow_excluded[[i]] <- nrow(excluded)
    included_summary <- ukb_summary(included) %>%
      rename(included = value)
    excluded_summary <- ukb_summary(excluded) %>%
      select(value) %>%
      rename(excluded = value)
    all_summary <- bind_cols(included_summary, excluded_summary) 
    all_summary$p_value <- str_pad(ifelse(ukb_p_value(all_summary, included, excluded) == 1,
                                          "1.", ukb_p_value(all_summary, included, excluded)), 
                                   6, "right", pad = "0")
    # write.csv(all_summary,
    #           file = paste(race_BFZ, "_", outcome, ".csv", sep = ""),
    #           row.names = FALSE)
    colnames(all_summary) <- c("Independent variables", "statistic", paste0(race_BFZ, "_included"),
                               paste0(race_BFZ, "_excluded"), paste0(race_BFZ, "_p_value"))
    if (i == 1) race_summary_all <- all_summary
    else race_summary_all <- full_join(race_summary_all, all_summary)
  }
  nrow_included_outcomes[[j]] <- nrow_included
  nrow_excluded_outcomes[[j]] <- nrow_excluded
  race_summary_outcomes[[j]] <- race_summary_all
}

covariate_list <- c("age", "sex", "bmi", "smoking_status", "alcohol_status", "townsend_index", "pa")
covariate_list_2 <- c("<i>Age (years)</i>", "<i>Sex, n (%)</i>", 
                      "<i>Body mass index (Kg/m<sup>2</sup>)</i>",
                      "<i>Smoking status, n (%)</i>", "<i>Alcohol status, n (%)</i>", 
                      "<i>Townsend index score</i>", 
                      "<i>Types of physical activity in last 4 weeks, n (%)</i>"
                      )
hux_table_2 <- ukb_hux_2(race_summary_outcomes, covariate_list, covariate_list_2, "glucose") %>%
  set_font_size(8)  %>%
  set_col_width(1, 0.11) %>%
  set_col_width(2, 0.1) %>%
  set_align(col = 3:11, value = "center") %>%
  set_valign(1, 1, "middle") %>%
  set_caption("<span style='color: black; font-weight: bold; font-size: 10pt'>Table 2: Participant characteristics: included vs excluded, blood glucose") %>%
  set_caption_pos("topleft") %>%
  add_footnote("<sup>a</sup>T-test for means, Mann Whitney for medians, and Fisher's Exact Test (<i>N</i> < 1000) and Chi-square (<i>N</i> ≥  1000) for categorical variables."
               , font_size = 7) %>%
  set_escape_contents(col = 1, value = FALSE) %>%
  set_escape_contents(row = 1:2, value = FALSE) %>%
  set_italic(col = 5, .[[5]] < 0.05) %>%
  set_italic(col = 8, .[[8]] < 0.05) %>%
  set_italic(col = 11, .[[11]] < 0.05) %>%
  set_bold(col = 5, .[[5]] < 0.05) %>%
  set_bold(col = 8, .[[8]] < 0.05) %>%
  set_bold(col = 11, .[[11]] < 0.05)
hux_table_2
```

```{r Table 2b, include = FALSE, eval = FALSE}
hux_table_2 <- hux_table_2 %>%
  set_width(0.9)  
quick_html(hux_table_2, file = "hux_table_2.html", open = FALSE)
webshot2::webshot(url = "hux_table_2.html", file = "hux_table_2.png", zoom = 5)
```
**Table 2** to be inserted here.

<br>

```{r Table 3, echo = FALSE, message = FALSE, eval = FALSE}
hux_table_3 <- ukb_hux_2(race_summary_outcomes, covariate_list, covariate_list_2, "urate") %>%
  set_font_size(8)  %>%
  set_col_width(1, 0.11) %>%
  set_col_width(2, 0.1) %>%
  set_align(col = 3:11, value = "center") %>%
  set_valign(1, 1, "middle") %>%
  set_caption("<span style='color: black; font-weight: bold; font-size: 10pt'>Table 3: Participant characteristics: included vs excluded, serum urate") %>%
  set_caption_pos("topleft") %>%
  add_footnote("<sup>a</sup>T-test for means, Mann Whitney for medians, and Fisher's Exact Test (<i>N</i> < 1000) and Chi-square (<i>N</i> ≥  1000) for categorical variables."
               , font_size = 7) %>%
  set_escape_contents(col = 1, value = FALSE) %>%
  set_escape_contents(row = 1:2, value = FALSE) %>%
  set_italic(col = 5, .[[5]] < 0.05) %>%
  set_italic(col = 8, .[[8]] < 0.05) %>%
  set_italic(col = 11, .[[11]] < 0.05) %>%
  set_bold(col = 5, .[[5]] < 0.05) %>%
  set_bold(col = 8, .[[8]] < 0.05) %>%
  set_bold(col = 11, .[[11]] < 0.05)
hux_table_3
```

```{r Table 3b, include = FALSE, eval = FALSE}
hux_table_3 <- hux_table_3 %>%
  set_width(0.9)  
quick_html(hux_table_3, file = "hux_table_3.html", open = FALSE)
webshot2::webshot(url = "hux_table_3.html", file = "hux_table_3.png", zoom = 5)
```
**Table 3** to be inserted here.

<br>

```{r Figure 4, include = FALSE, eval = FALSE}
# Quantile-quantile (QQ)  plots for normality for the outcomes (race-specific)
# Non-transformed
bd_BFZ_df <- read_rds("bd_BFZ_df.rds")
races <- c("White", "Black", "Asian")
png("QQ_outcomes.png", width=1500, height=1500, res=120)
par(mfrow = c(3,4))
for(i in seq_along(races)){
  race_BFZ <- races[[i]]
  bd_BFZ_race <- bd_BFZ_df %>% 
    filter(race == race_BFZ) %>%
    select(sbp, dbp, glucose, urate) %>%
    mutate(glucose = exp(glucose),
           urate = exp(urate),
           sbp = exp(sbp),
           dbp = exp(dbp)
    )
  ukb_qq_plots(bd_BFZ_race, race = race_BFZ)
}
dev.off()
```
**Figure 4. QQ plots for non-transformed outcomes.** to be inserted here.

<br>

```{r Figure 5, include = FALSE, eval = FALSE}
# Transformed
races <- c("White", "Black", "Asian")
png("QQ_log_outcomes.png", width=1500, height=1500, res=120)
par(mfrow = c(3,4))
for(i in seq_along(races)){
  race_BFZ <- races[[i]]
  bd_BFZ_race <- bd_BFZ_df %>% 
    filter(race == race_BFZ) %>%
    select(sbp, dbp, glucose, urate) 
  ukb_qq_plots(bd_BFZ_race, race = race_BFZ)
}
dev.off()
```
**Figure 5. QQ plots for log-transformed outcomes.** to be inserted here.

<br>

```{r Figure 6, include = FALSE, eval = FALSE}
# Transformed
for(i in seq_along(races)){
  race_BFZ <- races[[i]]
  
  bd_BFZ_race <- bd_BFZ_df %>% 
    filter(race == race_BFZ) %>%
    select(age:pa, sbp, dbp, glucose, urate) %>% 
    keep(is.numeric) 
  
  png(paste(race_BFZ, "_correlation_log_plot.png", sep =""), 
      width=1500, height=1500, res=120)
  pairs.panels(bd_BFZ_race)
  dev.off()
}
```
**Figure 6. An enhanced scatterplot matrix showing pairwise correlations between the variables.** to be inserted here.

<br>

```{r Figure 7, include = FALSE, eval = FALSE}
# Imputations (forwards, use imputed data set and transformed)
# Strip plots
bd_BFZ_df <- read_rds("bd_BFZ_df.rds")
races <- c("White", "Black", "Asian")
for(i in seq_along(races)){
  race_BFZ <- races[[i]]
  bd_imp <- bd_BFZ_df %>%
    filter(race == race_BFZ) %>%
    select(-f.eid, -I, -date, -race)
  tempData <- mice(bd_imp, m = 1, maxit = 10, seed = 7) # Impute missing data

  png(paste(race_BFZ, "_stripplot.png", sep =""),
      width=1500, height=1500, res=120)
  print(stripplot(tempData, pch = 20, cex = 1.2))
  dev.off()

  bd_imp <- as_tibble(complete(tempData, 1, include = FALSE)) %>%
    select(age:pa, glucose:dbp)
  write_rds(bd_imp, paste(race_BFZ, "_bd_imp.rds", sep =""))
}
```
**Figure 7. Strip plot showing imputed continuous covariates.** to be inserted here.

<br>

```{r Figure 8, include = FALSE, eval = FALSE}
# linearity assumptions
races <- c("White", "Black", "Asian")
outcomes <- c("glucose", "urate", "sbp", "dbp")
continuous_covariates <- c("age", "bmi", "townsend_index")

for(i in seq_along(races)){
  race_BFZ <- races[[i]]
  bd_imp <- read_rds(paste(race_BFZ, "_bd_imp.rds", sep =""))

  png(paste("Non_linearity_", race_BFZ, ".png", sep = ""),
      width = 1500, height = 1500, res = 120)
  par(mfrow = c(length(outcomes), length(continuous_covariates)))
  for(j in seq_along(outcomes)){
    outcome <- outcomes[[j]]
    for (k in seq_along(continuous_covariates)){
      continuous_covariate <- continuous_covariates[[k]]
      # ordinary least squares
      model_ols <- ols(bd_imp[[outcome]] ~ bd_imp[[continuous_covariate]])
      # restricted cubic splines, 5 knots
      model_rcs <- ols(bd_imp[[outcome]] ~ rcs(bd_imp[[continuous_covariate]],
                                               quantile(bd_imp[[continuous_covariate]],
                                                        c(0.05, 0.275, 0.5,
                                                          0.725, 0.95)
                                               )
      )
      )
      # anova test for non-linearity
      label_rcs <- paste("F-Test P for non-linearity = ",
                         round(anova(model_rcs, ss = FALSE)[" Nonlinear", "P"], 3),
                         sep = ""
      )
      # Likelihood ratio test
      label_lr <- paste("Likelihood Ratio Test P = ",
                         round(lrtest(model_ols, model_rcs)$stats["P"], 3),
                         sep = ""
      )
      label_rcs_lr <- paste(label_rcs, label_lr, sep = "; ")
      # make plot
      plot(x = bd_imp[[continuous_covariate]],
           y = bd_imp[[outcome]],
           xlab = ukb_label_names(continuous_covariate),
           ylab = paste(str_to_title(outcome), " (", ukb_label_names_2(outcome),
                        "); ", "log transformed ", sep = ""),
           main = paste(title_names(outcome), ": ",
                        title_names(continuous_covariate), sep = "")
      )
      lines(bd_imp[[continuous_covariate]],
            model_ols$fitted.values,
            col = "red", lwd = 2)
      points(bd_imp[[continuous_covariate]],
             model_rcs$fitted.values,
             col = "blue", pch = 15, cex = 1)
      text(min(bd_imp[[continuous_covariate]]) +
             (max(bd_imp[[continuous_covariate]]) -
                min(bd_imp[[continuous_covariate]])) * 0.5,
           max(bd_imp[[outcome]]),
           label_rcs_lr)
    }
  }
  dev.off()
}

# age + BMI to be included as splines for Whites
# ~ rcs(age, quantile(age, c(0.05, 0.275, 0.5, 0.725, 0.95))

```
**Figure 8. Non-linearity assessmnet using restricted cubic splines.** to be inserted here.

<br>

```{r Step 9, include = FALSE, eval = FALSE}
# Step 9: Prepare for GWAS
#-------------------------
races <- c("White", "Black", "Asian")
bd_BFZ_df <- read_rds("bd_BFZ_df.rds")
for (race in seq_along(races)){
  race_BFZ <- races[[race]]
  bd_imp <- bd_BFZ_df %>% 
    filter(race == race_BFZ) %>%
    select(-f.eid, -I, -date, -race)
  # md.pattern(bd_imp)   # Extent of missing data
  tempData <- mice(bd_imp, m = 1, maxit = 10, seed = 7) # Impute missing data 
  
  # with the continuous outcomes of age and sex as restricted cubic splines
  bd_imp <- as_tibble(complete(tempData, 1, include = FALSE)) %>%
    select(age:C10) %>%
    model.matrix(~ rcs(age, quantile(age, c(0.05, 0.275, 0.5, 0.725, 0.95)))
                 + sex + 
                   rcs(bmi, quantile(bmi, c(0.05, 0.275, 0.5, 0.725, 0.95))) 
                 + smoking_status + alcohol_status + townsend_index + pa + C1 + C2 
                 + C3 + C4 + C5 + C6 + C7 + C8 + C9 + C10, .
    ) %>%
    as_tibble() %>%
    select(-`(Intercept)`) 
  colnames(bd_imp)[1:9] <- c("age1", "age2", "age3", "age4", "sexMale", "bmi1", 
                             "bmi2", "bmi3", "bmi4")
    
  # to matrix format
  categorical <- c("sexMale", "smoking_statusPrevious", "smoking_statusCurrent",
                   "alcohol_statusPrevious", "alcohol_statusCurrent", "palight",
                   "paheavy")
  
  bd_imp[categorical] <- bd_imp %>% 
    select(all_of(categorical)) %>% 
    ukb_recode_snptest() 
  
  bd_GWAS <- bd_BFZ_df %>% 
    filter(race == race_BFZ) %>%
    select(f.eid, glucose:dbp) %>%
    bind_cols(bd_imp) %>% 
    relocate(glucose:dbp, .after = last_col()) %>% 
    rename(ID_1 = f.eid)  # to enable a mutating join later
  # outcomes at the end for SNPtest
  
  #Load sample file with ID1, ID2 and 'missing' columns
  sample_file <- as_tibble(fread("SAMPLE.csv")) #Load file with phenotypes
  sample_file <- left_join(sample_file, bd_GWAS)
  write.table(sample_file, file = paste("sample_", race_BFZ, ".sample", sep = ""), 
              sep = " ", row.names = FALSE)
}
# Manually edit the second row of the sample files to include D (discrete covariate),
# C (continuous covariate), B (binary outcome) and P (continuous outcome). 
# Or paste this in the second row: 0 0 0 C C C C D C C C C D D D D C D D C C C C C C C C C C P P P P
# Also replace any '"' with ''.
```

```{r Step 10, include = FALSE, eval = FALSE}
#Step 10: Running GWAS
#--------------------
# In an R environment, prepare linux scripts for transferring to server 
# using whites and sbp as an example. 

race <- "white"
outcome <- "sbp"
# with rcs (for whites)
for(i in 1:22){
  text <- paste("./snptest_v2.5.6 -data ukb_imp_chr", i, "_v3.bgen sample_", race, ".sample -pheno ", outcome, " -frequentist 1 -method expected -cov_names age1 age2 age3 age4 sexMale bmi1 bmi2 bmi3 bmi4 smoking_statusPrevious smoking_statusCurrent alcohol_statusPrevious alcohol_statusCurrent townsend_index palight paheavy C1 C2 C3 C4 C5 C6 C7 C8 C9 C10 -o UKBB_", race, "_", outcome ,"_chr",i, "_add.txt &> stdout", i, race, outcome, ".txt", sep="")
  cat(text, file = paste("SNPTest_", i, race, outcome,".sh", sep=""))
}
# without rcs (e.g. for blacks and asians), only include age1 and bmi1 as covariates for age and bmi
# i.e. remove age2, age3, age4, bmi2, bmi3 and bmi4 from the above script.

# # Run in Linux after transferring scripts
# race="white"
# outcome="sbp"
# for i in {1..22}; do
# chmod u+x SNPTest_${i}_*_${race}_${outcome}.sh
# done 
#
# for i in {1..22}; do
# ./SNPTest_${i}_*_${race}_${outcome}.sh &
# done
#
# # Run in Linux after the above SNPTest scripts have completed running
# for i in {1..22}; do
# sed '1,12d;$d' UKBB_${race}_${outcome}_chr${i}_add.txt > UKBB_${race}_${outcome}_chr0${i}.txt
# done
# cat UKBB_${race}_${outcome}_chr0*.txt > UKBB_${race}_${outcome}_chr1_22.txt
# 
# echo "SNP CHR BP P alleleA alleleB info all_AA all_AB all_BB all_total all_maf missing_data_proportion frequentist_add_beta_1 frequentist_add_se_1 hwe" > UKBB_${race}_${outcome}_Manhattan_temp.txt
# awk '{print $2,$3,$4,$22,$5,$6,$9,$14,$15,$16,$18,$19,$20,$24,$25,$21}' UKBB_${race}_${outcome}_chr1_22.txt >> UKBB_${race}_${outcome}_Manhattan_temp.txt
# awk '$4!="NA"' UKBB_${race}_${outcome}_Manhattan_temp.txt > UKBB_${race}_${outcome}_Manhattan.txt
# awk '$12>=0.01' UKBB_${race}_${outcome}_Manhattan.txt > UKBB_${race}_${outcome}_Manhattan_MAF.txt
# awk '$13<=0.05' UKBB_${race}_${outcome}_Manhattan_MAF.txt > UKBB_${race}_${outcome}_Manhattan_missingness.txt
# awk '$16>0.000001' UKBB_${race}_${outcome}_Manhattan_missingness.txt > UKBB_${race}_${outcome}_Manhattan_hwe.txt
# echo "SNP CHR BP P alleleA alleleB info all_AA all_AB all_BB all_total all_maf missing_data_proportion frequentist_add_beta_1 frequentist_add_se_1 hwe" > UKBB_${race}_${outcome}_Manhattan_info.txt
# awk '$7>0.4' UKBB_${race}_${outcome}_Manhattan_hwe.txt >> UKBB_${race}_${outcome}_Manhattan_info.txt
# 
# wc -l UKBB_${race}_${outcome}_Manhattan.txt > UKBB_${race}_${outcome}_included_SNPs.txt
# wc -l UKBB_${race}_${outcome}_Manhattan_MAF.txt >> UKBB_${race}_${outcome}_included_SNPs.txt
# wc -l UKBB_${race}_${outcome}_Manhattan_missingness.txt >> UKBB_${race}_${outcome}_included_SNPs.txt
# wc -l UKBB_${race}_${outcome}_Manhattan_hwe.txt >> UKBB_${race}_${outcome}_included_SNPs.txt
# wc -l UKBB_${race}_${outcome}_Manhattan_info.txt >> UKBB_${race}_${outcome}_included_SNPs.txt
# # _Manhattan_info.txt count includes the header.

# After downloading GWAS results from server (results placed in the R project directory)
results <- as_tibble(fread(paste("UKBB_", race, "_", outcome, 
                                 "_Manhattan_info.txt", sep = "")))  
results2 <- results %>%
  arrange(P)
write.csv(results2[results2$P < 1e-5,],  
          file = paste("UKBB_", race, "_", outcome, ".csv", sep = ""), 
          row.names = FALSE)
genome_significant <- nrow(results2[results2$P < 5e-8,])
nominally_significant <- nrow(results2[results2$P < 1e-5,])
analyzed_snps <- nrow(results)

# Manhattan plot
ukb_manhattan_plots(results, race, outcome)

# Quantile-quantile (QQ) plot
ukb_qq_gwas_plots(results, race, outcome)

# Genomic inflation factor (lambda)
lambda_white <- ukb_lambda(results)

# Getting genes and functional significance for nominally significant hits
races <- c("White", "Black", "Asian")
outcomes <- c("glucose", "urate", "sbp", "dbp")
for(i in seq_along(races)){
  race <- races[[i]]
  for(j in seq_along(outcomes)){
    outcome <- outcomes[[j]]
    ukb_dbsnp(race, outcome)
  }
}
```

```{r Step 10b, include = FALSE, eval = FALSE}
# Step 10b: Annotating Manhattan plots (example for "Whites" and "sbp")
# ------------------------------------
outcomes <- "glucose"
n_digits <- 3
race <- races[[1]]
outcome <- outcomes[[1]]
image_png <- image_read(paste0("UKBB_", race, "_", outcome, "_manhattan.png")) %>%
  ukb_image_annotate("G6PC2", 295, 330, bold = TRUE, decoration = "underline") %>%
  ukb_image_annotate("ABCB11", 295, 383) %>%
  ukb_image_annotate("EIF5A2", 400, 270) %>%
  ukb_image_annotate("SLC2A2", 405, 291, bold = TRUE, decoration = "underline") %>%
  ukb_image_annotate("TARS1-DT", 510, 395) %>%
  ukb_image_annotate("CAMK2B", 680, 283) %>%
  ukb_image_annotate("YKT6", 722, 415) %>%
  ukb_image_annotate("GCK", 673, 418, bold = TRUE, decoration = "underline") %>%
  ukb_image_annotate("SLC30A8", 782, 385, bold = TRUE, decoration = "underline") %>%
  ukb_image_annotate("TCF7L2", 920, 85, bold = TRUE, decoration = "underline") %>%
  ukb_image_annotate("   CCND2\nCCND2-AS1", 980, 380) %>%
  ukb_image_annotate(paste0("λ = ", 
                            str_pad(round(lambda, n_digits),
                                    n_digits + 1 + str_count(lambda, "[.]"), 
                                    "right", 
                                    pad = "0")
                            ), 
                     1375, 85, boxcolor = "grey", style = "normal") 
image_write(image_png, path = paste0("UKBB_", race, "_", outcome, "_manhattan_annotated.png"), format = "png")
```
**Figure 9. Manhattan plots.** to be inserted here.

<br>

\newpage
#### **References**

